### ISSUE TITLE
[Bug]: Why are the vLLM and Hugging Face Transformers inference results inconsistent?

### ISSUE THREAD

**Issue by Molasse** [vllm-project/vllm#12343]
### My current environment
<details>
  <summary>My current environment</summary>
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4

Python version: 3.10.12
Python platform: Linux-5.15.0-124-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 PCIe
GPU 1: NVIDIA H100 PCIe

Nvidia driver version: 535.183.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU Architecture: x86_64

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-ml-py==12.560.30
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pyzmq==26.2.0
[pip3] torch==2.5.1
[pip3] torchvision==0.20.1
[pip3] transformers==4.47.1
[pip3] triton==3.1.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.6.post1
vLLM Build Flags:

NCCL_SOCKET_IFNAME=bond0
CUDA_HOME=/usr/local/cuda-12.5
CUDA_MODULE_LOADING=LAZY
</details>

### 🐛 Describe the bug

Hello Team,
I recently conducted inference using both vLLM and Hugging Face Transformers with the same model and input.
However, I noticed that the output results were inconsistent between the two frameworks.

Here are my questions regarding this issue:  
1. What is the root cause of the inconsistency?  
2. How can I align the inference results between the two frameworks to make them consistent?

I look forward to your response. Thank you!

Below are the experimental settings I have tried in an attempt to align the results, but the outputs are still inconsistent. I have also provided the code for reproduction.

1. Model: Phi3-mini-3.8B
2. Precision: BF16
3. Identical versions of PyTorch and FlashAttention2
4. No sampling strategy
5. Single GPU execution, no model parallelism
6. No batch (batch size = 1)
7. Inference is performed using the `generate` method from vLLM and Hugging Face on the same input question

```python
import os
from vllm import LLM, SamplingParams
from transformers import AutoModelForCausalLM, AutoTokenizer

os.environ["CUDA_VISIBLE_DEVICES"] = "0"
model_path = 'microsoft/Phi-3-mini-4k-instruct'

tokenizer = AutoTokenizer.from_pretrained(model_path)

os.environ["VLLM_ATTENTION_BACKEND"] = "FLASH_ATTN"
llm = LLM(
    model=model_path,
    dtype='bfloat16',
    tensor_parallel_size=1,
    gpu_memory_utilization = 0.7,
    max_num_seqs = 1,
    disable_log_stats = True,
)

model = AutoModelForCausalLM.from_pretrained(
    model_path,
    device_map="cuda:0",
    torch_dtype="bfloat16",
    trust_remote_code=True,
    attn_implementation='flash_attention_2',
)
model.eval()

sampling_params = SamplingParams(temperature=0, top_k=1, max_tokens=1)

template = 'I will provide a movie review. Your task is to determine whether the sentiment of the review is positive or negative. Please respond with "Y" if the sentiment is positive, and "N" if the sentiment is negative. Do not include any explanations, just the single letter "Y" or "N".\n\nHere is the review:'
imdb_texts = ["I watched this movie because I like Nicolas Cage and well, I found it strange and completely pointless... so I decided to poke around a little bit and got my hands on the 70s copy of it. Wow. what a difference. The original one was way better. I'd like you all to know it did originally actually make a statement, it's existence did have a purpose. It was really the Christian public expressing their fear of paganism. If you dig deeper into it it also makes comments on life but I don't want to go into details, just, simply put, if you were disappointed and you'd like to know what it SHOULD look like, feel free to watch the 70s version, a little dated, but A lot better.",
 'Whether it\'s a good movie or not, films of this kind has to be made, i think. It remembers me of "I love Huckabees", a overwhelmingly puzzling movie with Isabelle Huppert being sodomized by a young American in a mud pond, in a merry sadistic-masochist way (??!!!!). I hope the director will go on stepping across the border, as though i felt the choices Vincenzo Natali made, were not always subtle (some of the scenes were unhappily kind of "tarte-à-la-crème", like a childish slapstick), speaking about script and cinematography\x85 The color of "Cube" was black, "Nothing" is white, more cheerful, surprisingly, than the former films of Natali.',
 "Weak, fast and multicolor,this is the Valvoline's movie in fact you can see always this brand of oil in a lot of scene. The real protagonist are the cars,weak performances of Cage and Duvall. A intresting Angelina Jolie is a unlikely mechanic. For the lovers of dream car(LAMBORGHINI and FERRARI over all).",
 'William S. Hart (as Jim Treen), the most eligible bachelor in Canyon City, is finally getting hitched, to pretty blonde waitress Leona Hutton (as Molly Stewart). His fiancée doesn\'t know it, but Mr. Hart is secretly the western town\'s "Most Wanted" bandit. However, Hart is planning to go straight, due to his marriage plans. Unfortunately, Ms. Hutton discovers Hart\'s secret stash, whilst cleaning up his untidy cabin; so, she calls off the wedding. Next, Hutton succumbs to the charms of mining swindler Frank Borzage (as W. Sloane Carey).<br /><br />Serviceable entertainment from superstar Hart; he was ranked no less than #1 at the box office, by Quigley Publications, for the years 1915 and 1916 (ahead of Mary Pickford). The principles perform capably. Later on, Frank Borzage was quite a director; and Leona Hutton, a suicide... <br /><br />**** A Knight of the Trails (8/20/15) William S. Hart ~ William S. Hart, Leona Hutton, Frank Borzage',
 "Henri Verneuil represented the commercial cinema in France from 1960-1980. Always strong at the box-office, and usually telling dramatic and suspenseful tales of casino robberies, mafia score-settling and World War II battles, Verneuil could be counted on to give us two solid hours of entertainment on Saturday night. He worked with the cream of the male actors of his day: Gabin, Belmondo, Fernandel, Delon, Sharif, Anthony Quinn. I... comme Icare is the only time he directed Yves Montand. It's an oddly static film, taking place mainly in offices and conference rooms, containing not one chase scene and hardly any violence.<br /><br />Montand gives a good performance, if somewhat dry, and he is well supported by the other actors. I couldn't help wondering what Costa-Gavras could have done with this story, on the basis of Z (the Lambrakis assassination) and L'aveu (the torture of Artur London in Czechoslovakia by Stalinists).",
 "This movie was packed pull of endless surprises! Just when you thought it couldn't get worse, they added more joints and more pink fuzzy-lined vans with raunchy sex scenes. As you can guess, I was a victim of the original version. We were tricked into watching it thinking it was Supervan, the host box which promised lasars, jail breaks, and much more. Who would have thought a Dollar Store Christmas present could have been so much fun!",
 "This is a very difficult movie, and it's almost impossible to get a handle on what's going on. At first it seems to be a rather pedestrian movie about a guy (Trelkovsky) who needs an apartment and rather crassly invites himself into one when the current tenant (a woman) commits suicide. Then the twists and turns start. Are the neighbors trying to kill him? And why are the dead tenant's clothes turning up in the apartment? One wonders, finally, if Trelkovksy _is_ the prior tenant. <br /><br />SPOILER SPOILER SPOILER SPOILER SPOILER SPOILER <br /><br />One of the tricks Polanski pulls on us is to lie to us. We assume when we see things from the point of view of a character that we see things as the character does and that there may be distortions of reality. We assume when the camera is showing us things from its omniscient point of view that we see actuality - but Polanski has the camera lie to us.",
 'Two houses, one street, one phone booth, one car, a girl next door, a boy next door and a zombie. This list of ingredients should suffice for a great horror movie. All you need is some blue light, ambient music and...done. Not in the hands of Dutch director van Rouveroy though! <br /><br />I like to organize "bad movie evenings" from time to time. The concept is really simple: get some booze, get some film-loving friends, and immerse yourself in the worst cinema can offer. For such an evening this peace of filth is one of the best. Laughs guaranteed!<br /><br />The bizarre thing is, van Rouveroy is still defending her film as if it were a great achievement. To be a witness to this you\'ll have to listen to the DVD\'s commentary track. Again: disbelieve and laughs guaranteed!',
 'Four teenage girlfriends drive to Fort Laurdale for spring break.Unfortunately they get a flat tire in Medley,Georgia and one of the girls witnesses a brutal murder deep in the woods.The local sheriff is behind the crime and the nightmare begins..."Shallow Grave" is a pleasant low-budget surprise.The cast is likable enough,the direction is steady and the violence is particularly nasty and misogynistic.Especially the second murder is pretty grim.The murderous sheriff isn\'t one-dimensional character-in a couple of scenes it seems that he feels remorse for what he\'s done.The subplot involving the two boys they meet in the diner goes nowhere,but the stalking scenes in the woods are tense and exciting.7 out of 10.']
for text in imdb_texts:
    messages = [ 
        {"role": "user", "content": f"{template}{text}"}, 
    ] 
    input_ids = tokenizer.apply_chat_template(messages, tokenize=True, add_generation_prompt=True, return_tensors="pt").to("cuda:0")
    vllm_output = llm.generate(prompt_token_ids=input_ids.tolist(), sampling_params= sampling_params, use_tqdm = False)[0].outputs[0].text.strip()
    hf_output = model.generate(input_ids, max_new_tokens=1, temperature=0, top_k=1, do_sample=False)
    hf_output = tokenizer.decode(hf_output[0], skip_special_tokens=False)[-1]
    if vllm_output != hf_output:
        print(vllm_output, hf_output)
```
The output of the above code is as follows:
N Y
Y N
N Y
Y N
Y N
Y N
N Y
N Y
Y N

The expectation is exactly the same.

**Comment by DarkLight1337** [comment]
How different are the results? Some minor differences are expected because of the kernels that are used in vLLM.

**Comment by Molasse** [comment]
@DarkLight1337 
First of all, thank you for your reply.

The inconsistency rate is about 4 to 5 percent.
Since my topic is classification task, which is sensitive to accuracy, I hope the results can be consistent.

> How different are the results? Some minor differences are expected because of precision errors in the kernels that are used in vLLM.

I have used the same attention backend and version. What part of vLLM do you mean by kernel?

**Comment by DarkLight1337** [comment]
I mean the custom ops under `csrc/` directory. vLLM uses a fork of FlashAttention (`vllm_flash_attn`) which may yield different results. Can you try running this in full precision? It might improve the consistency.

**Comment by hmellor** [comment]
Is it any better if you use `generation_config="auto"`? This should use whatever default sampling params come with the model (I haven't checked, but you might not have manually set them all).

**Comment by Molasse** [comment]
> Is it any better if you use `generation_config="auto"`? This should use whatever default sampling params come with the model (I haven't checked, but you might not have manually set them all).

Let me check and try

**Comment by Molasse** [comment]
> I mean the custom ops under `csrc/` directory. In addition to that, vLLM uses a fork of FlashAttention (`vllm_flash_attn`) which may yield different results. Can you try running this in full precision? It might improve the consistency.

@DarkLight1337 
When I used full precision, the inconsistency rate was reduced from 4.7% to 0.00005%. It's a lot better, but it also doubles the inference time.
In addition, I have tried to align the FlashAttention version, the FlashAttention installed by pip should be the same as vllm, which is 2.6.2.

```
xxxx/lib/python3.10/site-packages/vllm/vllm_flash_attn$ cat __init__.py
__version__ = "2.6.2"

# Use relative import to support build-from-source installation in vLLM
from .flash_attn_interface import (
    flash_attn_func,
    flash_attn_varlen_func,
    flash_attn_with_kvcache,
)
```

**Comment by neilbhutada** [comment]
In your code I can see that you used the **apply_chat_template** function; however, I think a **cleaner way** of using a chat template is using `llm.chat` instead of `llm.generate`.
